services/device/geolocation/geoclue_location_provider.cc

```diff
@@ -0,0 +1,414 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/services/device/geolocation/geoclue_location_provider.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/barrier_callback.h"
+#include "base/feature_list.h"
+#include "base/functional/bind.h"
+#include "base/functional/callback_forward.h"
+#include "base/functional/callback_helpers.h"
+#include "base/location.h"
+#include "base/memory/weak_ptr.h"
+#include "base/ranges/algorithm.h"
+#include "base/time/time.h"
+#include "components/dbus/thread_linux/dbus_thread_linux.h"
+#include "dbus/property.h"
+#include "services/device/public/cpp/device_features.h"
+#include "services/device/public/cpp/geolocation/geolocation_manager.h"
+#include "services/device/public/cpp/geolocation/geoposition.h"
+
+namespace device {
+
+namespace {
+
+constexpr char kServiceName[] = "org.freedesktop.GeoClue2";
+constexpr char kLocationInterfaceName[] = "org.freedesktop.GeoClue2.Location";
+constexpr char kClientInterfaceName[] = "org.freedesktop.GeoClue2.Client";
+constexpr char kManagerInterfaceName[] = "org.freedesktop.GeoClue2.Manager";
+constexpr char kManagerObjectPath[] = "/org/freedesktop/GeoClue2/Manager";
+constexpr char kBraveDesktopId[] = "com.brave.Browser";
+
+// This corresponds to the GeoClue2 AccuracyLevel: Neighborhood
+constexpr uint32_t kGeoClueAccuracyLow = 5;
+
+// This corresponds to the GeoClue2 AccuracyLevel: Exact
+constexpr uint32_t kGeoClueAccuracyHigh = 8;
+
+mojom::Geoposition GetErrorPosition() {
+  mojom::Geoposition response;
+  response.error_code = mojom::Geoposition_ErrorCode::POSITION_UNAVAILABLE;
+  response.error_message = "Unable to create instance of Geolocation API";
+  return response;
+}
+
+// Note: This method blocks because the call to NewSystemProvider is not
+// asynchronous. Fortunately it happens in the background, so this is safe. As
+// it's pretty unlikely that the presence of the GeoClue2 service changes in a
+// browser session, we cache the result here, so we only calculate this once.
+//
+// The easiest way to determine if a DBus service exists is to try and call a
+// method on it, and see if it fails. For this, I chose
+// |GeoClue2.Manager.GetClient| (this is cached on the GeoClue2 side, so it will
+// be the same client we get when we start our service).
+bool GeoClueAvailable() {
+  static bool checked = false;
```

Comment:
Why are these static?

Reply:
Cool, that makes sense (and good to know!). I've removed the caching for now, as this isn't called that frequently (~1 per site instance with geolocation access).